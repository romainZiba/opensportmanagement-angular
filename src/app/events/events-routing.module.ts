import { AuthGuard } from './../auth/guards/auth.guard';
import { NgModule } from '@angular/core';
import { RouterModule, Routes } from '@angular/router';
import { EventsPageComponent } from './pages/list/events-page.component';
import { NewEventPageComponent } from './pages/new-event/new-event-page.component';

const routes: Routes = [
  {
    path: '',
    component: EventsPageComponent,
    canActivate: [AuthGuard]
  },
  {
    path: 'new',
    component: NewEventPageComponent,
    canActivate: [AuthGuard]
  }
];

@NgModule({
  imports: [RouterModule.forChild(routes)],
  exports: [RouterModule]
})
export class EventsRoutingModule {}
