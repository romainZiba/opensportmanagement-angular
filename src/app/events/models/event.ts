export interface Event {
  _id: number;
  name: string;
  fromDateTime: string;
  toDateTime: string;
  placeId: number;
  placeName: string;
  presentMembers: any[];
  absentMembers: any[];
  waitingMembers: any[];
  openForRegistration: boolean;
  cancelled: boolean;
  teamId: number;
  localTeamName?: any;
  visitorTeamName?: any;
  localTeamImgUrl?: any;
  visitorTeamImgUrl?: any;
  visitorTeamScore?: any;
  localTeamScore?: any;
  done?: any;
}

export class EventCreateUpdate {
  teamId: number;
  name: string;
  fromDate: string;
  toDate: string;
  fromTime: string;
  toTime: string;
  placeId: number;
  isRecurrent: boolean;
  recurrenceDays?: string[];
  opponentId?: number;
  championshipId?: number;
  type: EventType;
}

export enum EventType {
  TRAINING = "TRAINING",
  MATCH = "MATCH",
  OTHER = "OTHER"
}
