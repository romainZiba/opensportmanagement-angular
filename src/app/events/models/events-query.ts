export interface EventsQuery {
  teamId: number,
  page: number,
  size: number,
  retrieveAll: boolean
}
