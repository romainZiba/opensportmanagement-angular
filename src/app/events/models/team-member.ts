export class TeamMember {
  username: string;
  firstName: string;
  lastName: string;
  roles: string[];
  licenceNumber: string;
  email: string;
  phoneNumber: string;
}
