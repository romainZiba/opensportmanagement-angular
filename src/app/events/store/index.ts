export * from "./events.actions";
export * from "./events.state";
