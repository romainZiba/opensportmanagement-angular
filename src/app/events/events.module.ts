import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';

import { EventsRoutingModule } from './events-routing.module';
import { SharedModule } from '../shared/shared.module';
import { NgxsModule } from '@ngxs/store';
import { EventsState } from './store';
import { EventListComponent } from './components/event-list/event-list.component';
import { EventsPageComponent } from './pages/list/events-page.component';
import { EventItemComponent } from './components/event-item/event-item.component';
import { EventService } from './services/event.service';
import { NewEventPageComponent } from './pages/new-event/new-event-page.component';
import { NgxsFormPluginModule } from '@ngxs/form-plugin';

@NgModule({
  imports: [
    CommonModule,
    SharedModule,
    EventsRoutingModule,
    NgxsModule.forFeature([EventsState]),
    NgxsFormPluginModule
  ],
  providers: [EventService],
  declarations: [EventItemComponent, EventListComponent, EventsPageComponent, NewEventPageComponent]
})
export class EventsModule {}
