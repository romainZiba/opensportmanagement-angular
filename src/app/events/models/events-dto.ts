import { Event } from './event';

export interface Embedded {
  eventDtoes: Event[];
}

export interface First {
  href: string;
}

export interface Self {
  href: string;
}

export interface Next {
  href: string;
}

export interface Last {
  href: string;
}

export interface Links {
  first: First;
  self: Self;
  next: Next;
  last: Last;
}

export interface Page {
  size: number;
  totalElements: number;
  totalPages: number;
  number: number;
}

export interface EventDtos {
  _embedded: Embedded;
  _links: Links;
  page: Page;
}
