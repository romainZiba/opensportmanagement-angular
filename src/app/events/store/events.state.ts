import { Router } from "@angular/router";
import { Action, Selector, State, StateContext } from "@ngxs/store";

import { NgxsEntityAdapter, NgxsEntityStateModel } from "../../shared/entity";
import { EventService } from "../services/event.service";
import * as eventActions from "./events.actions";
import { LoadEvents, LoadEventsSuccess } from "./events.actions";
import { asapScheduler, of } from "rxjs/index";
import { catchError, map } from "rxjs/operators";

import { Event } from "../models/event";

export class EventStateModel extends NgxsEntityStateModel<Event> {
  loading: boolean;
  loaded: boolean;
}

@State<EventStateModel>({
  name: "events",
  defaults: {
    ...EventStateModel.initialState(),
    loaded: false,
    loading: false
  }
})
export class EventsState {
  constructor(private eventService: EventService, private router: Router) {}

  @Selector()
  static getEventEntities(state: EventStateModel) {
    return state.entities;
  }

  @Selector()
  static getSelected(state: EventStateModel) {
    return state.selected;
  }

  @Selector()
  static getIds(state: EventStateModel) {
    return state.ids;
  }

  @Selector()
  static isLoading(state: EventStateModel) {
    return state.loading;
  }

  @Selector()
  static isLoaded(state: EventStateModel) {
    return state.loaded;
  }

  @Selector()
  static getAllEvents(state: EventStateModel) {
    return NgxsEntityAdapter.getItemsFromEntities(state.entities);
  }

  @Action(LoadEvents)
  loadEvents(
    { patchState, dispatch }: StateContext<EventStateModel>,
    { payload }: LoadEvents
  ) {
    patchState({
      loading: true
    });
    return this.eventService
      .getEvents(
        payload.teamId,
        payload.page,
        payload.size,
        payload.retrieveAll
      )
      .pipe(
        map(
          dtos =>
            asapScheduler.schedule(() => {
              const events =
                dtos.page.totalPages > 0 ? dtos._embedded.eventDtoes : [];
              dispatch(new eventActions.LoadEventsSuccess(events));
            }),
          catchError(error =>
            of(
              asapScheduler.schedule(() =>
                dispatch(new eventActions.LoadEventsFailed(error))
              )
            )
          )
        )
      );
  }

  @Action(LoadEventsSuccess)
  loadEventsSuccess(
    ctx: StateContext<EventStateModel>,
    { payload }: eventActions.LoadEventsSuccess
  ) {
    NgxsEntityAdapter.addAll(payload, ctx, "_id");
    ctx.patchState({
      loaded: true,
      loading: false
    });
  }
}
