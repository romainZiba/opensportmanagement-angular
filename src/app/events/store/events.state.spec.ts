import { async, TestBed } from '@angular/core/testing';
import { NgxsModule, Store } from '@ngxs/store';
import { LoadEvents, LoadEventsSuccess } from './events.actions';
import { of } from 'rxjs/observable/of';
import { Event } from '../models/event';
import { EventService } from '../services/event.service';
import { HttpClientTestingModule } from '@angular/common/http/testing';
import { EventsState, EventStateModel } from './events.state';
import { RouterTestingModule } from '@angular/router/testing';
import { EventDtos } from '../models/events-dto';

// Return a fixed timestamp when moment().format() is called
jest.mock('moment', () => () => ({ format: () => '2018–11–03' }));

describe('Events', () => {
  let store: Store;
  let service: EventService;

  const event1: Event = {
    _id: 1,
    name: 'Event #1',
    fromDateTime: '',
    toDateTime: '',
    placeId: 1,
    placeName: 'place',
    presentMembers: [],
    absentMembers: [],
    waitingMembers: [],
    cancelled: false,
    openForRegistration: false,
    teamId: 1
  };

  const event2: Event = {
    _id: 2,
    name: 'Event #2',
    fromDateTime: '',
    toDateTime: '',
    placeId: 1,
    placeName: 'place',
    presentMembers: [],
    absentMembers: [],
    waitingMembers: [],
    cancelled: false,
    openForRegistration: false,
    teamId: 1
  };

  const events: Event[] = [event1, event2];
  const eventEntities = {
    1: event1,
    2: event2
  };

  const dtos: EventDtos = {
    _embedded: [],
    _links: [],
    page: 1
  };

  const selected = event1._id;
  const eventsState: EventStateModel = {
    entities: eventEntities,
    selected,
    ids: events.map(event => event._id),
    loaded: true,
    loading: false
  };

  beforeEach(async () => {
    TestBed.configureTestingModule({
      imports: [HttpClientTestingModule, RouterTestingModule, NgxsModule.forRoot([EventsState])],
      providers: [EventService]
    });

    service = TestBed.get(EventService);
    store = TestBed.get(Store);

    spyOn(service, 'getEvents').and.returnValue(of(dtos));
  });

  describe('Selector', () => {
    it('should select events entities', () => {
      expect(EventsState.getEventEntities(eventsState)).toEqual(eventEntities);
    });

    it('should select all events', () => {
      expect(EventsState.getAllEvents(eventsState)).toEqual(events);
    });

    it('should select all ids', () => {
      expect(EventsState.getIds(eventsState)).toEqual(events.map(e => e._id));
    });

    it('should select the selected event', () => {
      expect(EventsState.getSelected(eventsState)).toEqual(event1._id);
    });

    it('should tell if events are loading', () => {
      expect(EventsState.isLoading(eventsState)).toEqual(false);
    });

    it('should tell if events are loaded', () => {
      expect(EventsState.isLoaded(eventsState)).toEqual(true);
    });
  });

  describe('Actions', function() {
    describe('Load events', function() {
      it('should change loading to true', async(() => {
        const eventsQuery = { teamId: 1, page: 0, size: 20, retrieveAll: true };
        store.dispatch(new LoadEvents(eventsQuery));
        store.selectOnce(state => state.events.loading).subscribe(loading => {
          expect(loading).toBe(true);
        });
        expect(service.getEvents).toHaveBeenCalledWith(
          eventsQuery.teamId,
          eventsQuery.page,
          eventsQuery.size,
          eventsQuery.retrieveAll
        );
      }));
    });

    describe('Load events success', function() {
      it('should change state to reflect events loaded', async(() => {
        store.dispatch(new LoadEventsSuccess(events));
        store.selectOnce(state => state.events).subscribe(state => {
          expect(state).toMatchSnapshot();
        });
      }));
    });
  });
});
