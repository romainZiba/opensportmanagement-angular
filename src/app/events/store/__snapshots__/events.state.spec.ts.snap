// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Events Actions Load events success should change state to reflect events loaded 1`] = `
Object {
  "entities": Object {
    "1": Object {
      "_id": 1,
      "absentMembers": Array [],
      "cancelled": false,
      "fromDateTime": "",
      "name": "Event #1",
      "openForRegistration": false,
      "placeId": 1,
      "placeName": "place",
      "presentMembers": Array [],
      "teamId": 1,
      "toDateTime": "",
      "waitingMembers": Array [],
    },
    "2": Object {
      "_id": 2,
      "absentMembers": Array [],
      "cancelled": false,
      "fromDateTime": "",
      "name": "Event #2",
      "openForRegistration": false,
      "placeId": 1,
      "placeName": "place",
      "presentMembers": Array [],
      "teamId": 1,
      "toDateTime": "",
      "waitingMembers": Array [],
    },
  },
  "eventForm": Object {
    "dirty": false,
    "errors": Object {},
    "model": Object {
      "fromDate": "2018-11-03",
      "fromTime": "20:00",
      "isRecurrent": false,
      "name": "",
      "placeId": null,
      "toDate": "2018-11-03",
      "toTime": "22:30",
      "type": "MATCH",
    },
    "status": "",
  },
  "ids": Array [
    1,
    2,
  ],
  "loaded": true,
  "loading": false,
  "selected": null,
}
`;
